tests/Doctrine/Tests/DBAL/Functional/MasterSlaveConnectionTest.php

```diff
@@ -139,4 +143,15 @@ public function testMasterSlaveConnectionCloseAndReconnect()
         $conn->connect('master');
         $this->assertTrue($conn->isConnectedToMaster());
     }
+
+    public function testMasterSlaveFailSlave()
+    {
+        $params = $this->prepareConnectionParameters();
+        $invalidParams = $params['master'];
+        $invalidParams['password'] = 'MUUHAAAAHAAAA';
```

Comment:
You are right, this implementation is not complete, because cover too many cases(like bad auth params) so i think how to make it better ;)
Functionality i want is not retry on failed connection just try other slave or master because if slave is down or network between app server and database server is broken i want to fallback connection to other host